gph/bindings/ripser_bindings.cpp

```diff
@@ -15,85 +15,84 @@
 namespace py = pybind11;
 
 #if defined USE_COEFFICIENTS
-PYBIND11_MODULE(gph_ripser_coeff, m)
-{
+PYBIND11_MODULE(gph_ripser_coeff, m) {
 #else
-PYBIND11_MODULE(gph_ripser, m)
-{
+PYBIND11_MODULE(gph_ripser, m) {
 #endif
 
-    using namespace pybind11::literals;
-    m.doc() = "Ripser python interface";
+  using namespace pybind11::literals;
+  m.doc() = "Ripser python interface";
 
-    // Because `ripser` could have two different modules after compilation
-    // It's necessary to add `py::module_local()` to prevent following issue:
-    // ImportError: generic_type: type "ripserResults" is already registered!
-    // When same python module imports gtda_ripser and gtda_ripser_coeff
-    py::class_<ripserResults>(m, "ripserResults", py::module_local())
-        .def_readwrite("births_and_deaths_by_dim",
-                       &ripserResults::births_and_deaths_by_dim)
-        .def_readwrite("num_edges", &ripserResults::num_edges);
+  // Because `ripser` could have two different modules after compilation
+  // It's necessary to add `py::module_local()` to prevent following issue:
+  // ImportError: generic_type: type "ripserResults" is already registered!
+  // When same python module imports gtda_ripser and gtda_ripser_coeff
+  py::class_<ripserResults>(m, "ripserResults", py::module_local())
+      .def_readwrite("births_and_deaths_by_dim",
+                     &ripserResults::births_and_deaths_by_dim)
+      .def_readwrite("cocycles_by_dim", &ripserResults::cocycles_by_dim)
+      .def_readwrite("num_edges", &ripserResults::num_edges);
 
-    m.def(
-        "rips_dm",
-        [](py::array_t<value_t>& D, int N, int modulus, int dim_max,
-           float threshold, int num_threads) {
-            // Setup distance matrix and figure out threshold
-            auto D_ = static_cast<value_t*>(D.request().ptr);
-            std::vector<value_t> distances(D_, D_ + N);
-            compressed_lower_distance_matrix dist =
-                compressed_lower_distance_matrix(
-                    compressed_upper_distance_matrix(std::move(distances)));
+  m.def(
+      "rips_dm",
+      [](py::array_t<value_t>& D, int N, int modulus, int dim_max,
+         float threshold, int num_threads, bool do_cocycles) {
+        // Setup distance matrix and figure out threshold
+        auto D_ = static_cast<value_t*>(D.request().ptr);
+        std::vector<value_t> distances(D_, D_ + N);
+        compressed_lower_distance_matrix dist =
+            compressed_lower_distance_matrix(
+                compressed_upper_distance_matrix(std::move(distances)));
 
-            // TODO: This seems like a dummy parameter at the moment
-            float ratio = 1.0;
-            int num_edges = 0;
+        // TODO: This seems like a dummy parameter at the moment
+        float ratio = 1.0;
+        int num_edges = 0;
```

Comment:
Do we return `num_edges`?  It is such a simple operation and I'm not sure we want to waste computational time for it especially if it's not returned.

Reply:
We return it. Even in the C++ code it is not really costly. In my opinion we can keep it because it doesn't have a significant impact on run-time. But, in the Python interface we don't use it for the moment.

Reply:
I guess I would make the same point as you made for `ratio`.  I would slightly prefer if the bindings were really minimalistic and there was no computation done in them.

Reply:
You are maybe right. Then lets add it to the list of *minor* changes to add for release `0.2`